[hlsl-out] clear `named_expressions` inserted by duplicated blocks

teoxoy opened the pull request:
fixes #1972

jimblandy commented on src/back/hlsl/writer.rs:
```diff
@@ -1876,6 +1877,27 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
+                } => {
+                    self.named_expressions.remove(&handle);
+                }
+                _ => {}
```
Doesn't this need to recurse into sub-statements?

teoxoy commented:
Another way of doing this (which avoids the recursion) would be to use `indexmap::IndexMap` for `named_expressions` and use `.len()` + `.truncate(prev_len)`.

I'll push a commit that does this (which I can revert if we don't want to go this route).

teoxoy commented:
Seems like we need https://github.com/bluss/indexmap/issues/215 for that to work...

teoxoy commented:
I'll revert the last commit, adding `Arbitrary` support to `indexmap` also requires adding it to `hashbrown`. If we want to make it a newtype instead, it also gets really ugly since we need to expose all the methods of the inner `IndexMap` type.

cuviper commented:
> adding `Arbitrary` support to `indexmap` also requires adding it to `hashbrown`.

I'm not sure why you think so... bluss/indexmap#246

teoxoy commented:
@cuviper a PR, that's great! I was looking into deriving it (which would have required RawTable in hashbrown to also impl the trait).

teoxoy commented:
@cuviper do you have an ETA on that PR landing and possibly a new release?

cuviper commented:
Ah, deriving the fields directly would be *unlikely* to produce a coherent map. I just followed how each `Arbitrary` is implemented for a regular `HashMap`, with an iterator of `(K, V)` items to collect, or just `T` items for a set.

I went ahead and merged that on `master`. That's currently in a limbo 2.0.0 state though, which I should probably get around to publishing already, but I could also backport it to the 1.x series if you need that.

teoxoy commented:
If it's not too much trouble, it would be greatly appreciated :)

cuviper commented:
Done -- `indexmap 1.9.2`

ErichDonGubler commented:
@jimblandy: Were you still going to review this for another round? I'm thinking of trying to tackle this.

jimblandy commented:
Now that I actually understand what it's doing (from reviewing the switch PR), I was going to look it over again. Although, I'm pretty buried in #2075 right now.

ErichDonGubler reviewed (approved):
LGTM! Some non-blocking meta nits: 

* I sorta wish the justification for `indexmap` would land in the message of the commit introducing it, rather than just being GH discussion. That would make it easier for folks examining history. `git commit --fixup=amend…` is your friend!
* It's sorta odd to introduce a solution to the problem then immediately replace it in the same PR, but I don't see any issue with it.

teoxoy commented on tests/out/hlsl/control-flow.hlsl:
```diff
@@ -101,22 +101,36 @@ void main(uint3 global_id : SV_DispatchThreadID)
+        }
+        default: {
+            int i_2 = int(1);
+            break;
```
Yeah, we don't have any optimization passes. All the compilers naga targets have their own already.

teoxoy commented:
Thanks for the review!

> It's sorta odd to introduce a solution to the problem then immediately replace it in the same PR, but I don't see any issue with it.

I left all commits in for now because I wanted some feedback on both approaches and was planning to squash everything when rebasing (before merging).

> I sorta wish the justification for `indexmap` would land in the message of the commit introducing it, rather than just being GH discussion. That would make it easier for folks examining history. `git commit --fixup=amend…` is your friend!

Will do.

teoxoy commented:
Squashed and rebased. I had to remove the test since we removed support for fallthrough for the wgsl frontend in https://github.com/gfx-rs/naga/issues/2031.

ErichDonGubler reviewed (approved):
Still LGTM!

jimblandy reviewed (approved):
This looks great - much simpler than the original fix. It's a pity we can't test this.
